src/System.Management.Automation/engine/NativeCommandProcessor.cs

```diff
@@ -528,90 +541,431 @@ internal override void Complete()
+                        _path,
+                        originalException.Message);
+                    XmlException newException = new XmlException(
+                        message,
+                        originalException);
+
+                    ErrorRecord error = new ErrorRecord(
+                        newException,
+                        "ProcessStreamReader_CliXmlError",
+                        ErrorCategory.SyntaxError,
+                        _path);
+                    result.Add(new ProcessOutputObject(error, MinishellStream.Error));
+                }
+
+                return result;
+            }
+        }
+
+        private void InitOutputQueue()
+        {
+            //if output is redirected, start reading output of process in queue.
+            if (_nativeProcess.StartInfo.RedirectStandardOutput || _nativeProcess.StartInfo.RedirectStandardError)
+            {
+                lock (_sync)
+                {
+                    if (!_stopped)
+                    {
+                        _nativeProcessOutputQueue = new BlockingCollection<ProcessOutputObject>();
+                        // we don't assign the handler to anything, because it's used only for objects marshaling
+                        new ProcessOutputHandler(_nativeProcess, _nativeProcessOutputQueue);
+                    }
+                }
+            }
+        }
+
+        private ProcessOutputObject DequeueProcessOutput(bool blocking)
+        {
+            if (blocking)
+            {
+                // if adding was completed, there is no need to do a blocking Take(),
+                // if collection is empty
+                if (_nativeProcessOutputQueue.IsAddingCompleted)
+                {
+                    if (_nativeProcessOutputQueue.Count > 0)
+                    {
+                        return _nativeProcessOutputQueue.Take();
+                    }
                 }
-                finally
+                else
                 {
-                    if (background == false)
+                    try
+                    {
+                        return _nativeProcessOutputQueue.Take();
+                    }
+                    catch (InvalidOperationException)
                     {
-                        //Wait for process to exit
-                        _nativeProcess.WaitForExit();
+                        // It's a normal situation: another thread can mark collection as CompleteAdding
+                        // in a concurrent way and we will rise an exception in Take().
+                        // Although it's a normal situation it's not the most common path
+                        // and will be executed only on the race condtion case.
+                    }
+                }
 
-                        //Wait for input writer to finish.
-                        _inputWriter.Done();
+                // collection is empty or exception been raised
+                return null;
```

Comment:
> Bear in mind that this has nothing to do with try/catch blocks: **you only incur the cost when the actual exception is thrown.** You can use as many try/catch blocks as you want. Using exceptions gratuitously is where you lose performance.

This is from a MSDN article [Performance Tips and Tricks in .NET Applications](https://msdn.microsoft.com/en-us/library/ms973839.aspx?ppud=4). It indicates that try/catch block itself doesn't incur any performance overhead, as long as no exception is thrown.